imap/imapd.c

```diff
@@ -286,108 +291,174 @@ struct list_rock_recursivematch {
     struct list_entry *array;
 };
 
-/* CAPABILITIES are defined here, not including TLS/SASL ones,
-   and those that are configurable */
+/* CAPABILITIES are defined here */
 
 enum {
-    CAPA_PREAUTH = 0x1,
-    CAPA_POSTAUTH = 0x2
+    CAPA_PREAUTH     = (1<<0),   /* advertised pre-authentication             */
+    CAPA_POSTAUTH    = (1<<1),   /* advertised post-authentication            */
+    CAPA_REQD        = (1<<2),   /* required (can't be suppressed)            */
+    CAPA_STATE       = (1<<3),   /* depends on a server state variable        */
+    CAPA_CONFIG      = (1<<4),   /* depends on an imapd.conf config switch    */
+    CAPA_REVERSE     = (1<<5),   /* reverse state/switch logic                */
+    CAPA_COMPLEX     = (1<<6),   /* function to print "complex" capability    */
+    CAPA_VALUE       = (1<<7),   /* has a formatted value based on config opt */
+    CAPA_MULTI       = (1<<8),   /* has multiple string values                */
 };
 
+#define CAPA_OMNIAUTH  (CAPA_PREAUTH|CAPA_POSTAUTH)
+#define CAPA_REVCONFIG (CAPA_CONFIG|CAPA_REVERSE)
+
 struct capa_struct {
-    const char *str;
-    int mask;
+    const char *str;            /* capability name                            */
+    int mask;                   /* mask of CAPA_* flags                       */
+
+    /* availability/value(s) optionally based on one of the following:        */
+    union {
+        int *statep;            /* CAPA_STATE:   ptr to server state variable */
+        enum imapopt config;    /* CAPA_CONFIG:  config switch option name    */
+        void (*complex)(void);  /* CAPA_COMPLEX: function to print capablity  */
```

Comment:
Doesn't look like anything uses `complex` at the moment, is there something you anticipate will need it, or is this just general future-proofing?